eax/src/lib.rs

```diff
@@ -0,0 +1,286 @@
+        let h = Self::cmac_with_iv(&self.key, 1, associated_data);
+
+        // 3. enc ← CTR(M) using n as iv
+        let mut cipher = ctr::Ctr128::<Cipher>::from_block_cipher(Cipher::new(&self.key), &n);
+        cipher.apply_keystream(buffer);
+
+        // 4. c ← OMAC(2 || enc)
+        let c = Self::cmac_with_iv(&self.key, 2, buffer);
+
+        // 5. tag ← n ^ h ^ c
+        // (^ means xor)
+        Ok(n.zip(h, |a, b| a ^ b).zip(c, |a, b| a ^ b))
+    }
+
+    fn decrypt_in_place_detached(
+        &self,
+        nonce: &GenericArray<u8, Self::NonceSize>,
+        associated_data: &[u8],
+        buffer: &mut [u8],
+        tag: &Tag,
+    ) -> Result<(), Error> {
+        if buffer.len() as u64 > C_MAX || associated_data.len() as u64 > A_MAX {
+            return Err(Error);
+        }
+
+        // 1. n ← OMAC(0 || Nonce)
+        let n = Self::cmac_with_iv(&self.key, 0, nonce);
+
+        // 2. h ← OMAC(1 || associated data)
+        let h = Self::cmac_with_iv(&self.key, 1, associated_data);
+
+        // 4. c ← OMAC(2 || enc)
+        let c = Self::cmac_with_iv(&self.key, 2, buffer);
+
+        // 5. tag ← n ^ h ^ c
+        // (^ means xor)
+        let expected_tag = n.zip(h, |a, b| a ^ b).zip(c, |a, b| a ^ b);
+
+        let expected_tag = &expected_tag[..tag.len()];
+
+        // Check mac using secure comparison
+        use subtle::ConstantTimeEq;
+        if expected_tag.ct_eq(tag).unwrap_u8() == 1 {
+            // Decrypt
+            let mut cipher = ctr::Ctr128::<Cipher>::from_block_cipher(Cipher::new(&self.key), &n);
+            cipher.apply_keystream(buffer);
+            Ok(())
+        } else {
+            Err(Error)
+        }
+    }
+}
+
+impl<Cipher> Eax<Cipher>
+where
+    Cipher: BlockCipher<BlockSize = U16> + NewBlockCipher + Clone,
+    Cipher::ParBlocks: ArrayLength<Block<Cipher>>,
+{
+    /// CMAC/OMAC1
+    ///
+    /// To avoid constructing new buffers on the heap, an iv encoded into 16
+    /// bytes is prepended inside this function.
+    fn cmac_with_iv(
+        key: &GenericArray<u8, Cipher::KeySize>,
+        iv: u8,
+        data: &[u8],
+    ) -> GenericArray<u8, <Cmac<Cipher> as Mac>::OutputSize> {
+        let mut mac = Cmac::<Cipher>::new(key);
+        mac.update(&[0; 15]);
+        mac.update(&[iv]);
+        mac.update(data);
+
+        mac.finalize().into_bytes()
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    use quickcheck::quickcheck;
+    use std::prelude::v1::*;
+
+    #[test]
+    fn known_data() {
+        let mut data = [1, 2, 3, 4, 5];
+        let cipher = Aes128Eax::new(&[0; 16].into());
+        let mac = cipher
+            .encrypt_in_place_detached(&[0; 16].into(), &[1, 2, 3, 4], &mut data)
+            .unwrap();
+        assert_eq!(
+            mac.as_slice(),
+            &[232, 88, 147, 206, 130, 126, 14, 121, 62, 127, 33, 233, 239, 81, 51, 177]
+        );
+        assert_eq!(&data, &[182, 81, 68, 170, 62]);
+    }
+
+    quickcheck! {
```

Comment:
Do we want to property-based testing here? I think it will unnecessary make CI runs longer.